### DISCUSSION TITLE
Use curv in a headless server

### DISCUSSION THREAD

**Discussion by A-G-D** [curv3d/curv#131] [Q&A] [answered]
Is there a way to use curv (specifically to export a curv script into a PNG) in a headless server?
I'm using curv in Github-Actions to generate a PNG from a curv script, and here you can see the problem:
> `+ ./AppDir/usr/bin/curv -o curv.png -O xsize=512 -O ysize=512 /home/runner/work/curv/curv/icon.curv`
> `GLFW error 0x10008: X11: The DISPLAY environment variable is missing`
> `ABORT: GLFW init failed`

Setting `DISPLAY` variable does't solve the problem either:

> `+ export DISPLAY=:0`
> `+ DISPLAY=:0`
> `+ icon_filename=curv.png`
> `+ ./AppDir/usr/bin/curv -o curv.png -O xsize=512 -O ysize=512 /home/runner/work/curv/curv/icon.curv`
> `GLFW error 0x10008: X11: Failed to open display :0`
> `ABORT: GLFW init failed`

Any thoughts?

**Comment by doug-moen** [comment, marked as answer]
I know you can run Curv on Linux without a GPU if you use the LLVMPIPE software gpu driver, but you still need an X server (or a Wayland server). It's ultimately an issue with the GLFW library that I use. Other GLFW users want this feature, but it isn't available. Eg, see https://github.com/glfw/glfw/issues/1882

Maybe you can configure a dummy X server to run on your headless server? I have no experience with this, but there are some mentions on this on the web. Eg, https://techoverflow.net/2019/02/23/how-to-run-x-server-using-xserver-xorg-video-dummy-driver-on-ubuntu/

## Background Info
Curv uses the GPU to render a PNG. It accesses the GPU via OpenGL. To create the OpenGL context, it uses the GLFW library. Specifically, this code in `libcurv/viewer/viewer.cc`:
```
if (headless_) {
    glfwWindowHint(GLFW_VISIBLE, GL_FALSE);
}
glfwCreateWindow(window_size_.x, window_size_.y, appTitle, NULL, NULL);
```
I create an (invisible) window in headless mode because in GLFW, that's the only way to create a context.
> GLFW doesn't support creating contexts without an associated window. However, contexts with hidden windows can be created with the GLFW_VISIBLE window hint.
> https://www.glfw.org/docs/latest/context.html#context_offscreen

**Reply by A-G-D** [reply]
I tried the dummy server workaround that you linked and it worked, Thanks.
